### PR TITLE
feat(dev-exp): Split entity processing and storage processing stages for simple queries

### PR DESCRIPTION
**Pull request by enochtangg**
### Overview
With this change, we can run two different query pipelines for simple queries:
1. The existing one (coupled entity and storage processing)
2. The new one (decoupled entity and storage processing)

Every snuba query needs a query plan to run. The query plan also happens to store information about the storage processors that need to be applied. The problem today is that in order create a query plan, we need a query plan builder, and to get one of those, we need the entity. This is the relationship that couples entities and storages together.

To decouple them, the query plan and query plan builder classes needed to be modified. To support this, an existing file was duplicated and suffixed with `_new`, classes were modified and suffixed with `New`, and a new `EntityProcessingExecutor` class was introduced. See files:
* `snuba/datasets/plans/storage_plan_builder_new.py`

Lastly, the `EntityAndStoragePipelineStage` was divided into `EntityProcessingStage` and `StorageProcessingStage` which leverages the newly defined classes to handling their respective processing.

### Rollout Plan
When this PR is merged, nothing changes. We configure the rollout through the runtime config. First, use `try_new_query_pipeline` which samples a percentage a queries, dry runs it in the new pipeline, and compares the difference in SQL produced with the SQL from the existing pipeline. Once we confirm there are no differences, we can use `run_new_query_pipeline` to actually run the query using the new pipeline.

```
SET try_new_query_pipeline = 0.01 # 1% of queries try using the new pipeline
SET run_new_query_pipeline = 0.01 # 1% of queries go through new pipeline
```

Monitor in sentry and DD, and continue increasing `run_new_query_pipeline` until we hit 100%. 
After support is added for composite queries, we can remove the old class names and files.

### CODE REVIEW

**Comment by volokluev** [comment]
So what happens if we get a composite query?

**Comment by enochtangg** [comment]
> So what happens if we get a composite query?

@volokluev It will run the existing pipeline stage instead: https://github.com/getsentry/snuba/pull/5699/files#diff-34537e7c1018d32db9667291e64b7c57da4c6e148725b1e97864e646c588f4e6R59-R62

**Review by evanh — COMMENTED** [review]
I'm a little confused at this: the Entity still build the storage query plan builder. And that builder is still run in the EntityProcessingStage.

This doesn't seem to accomplish the goal laid out in the title of the PR. Is this a first step? If so, what's the next step?

File: snuba/datasets/plans/storage_plan_builder_new.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+        partition_key_column_name: Optional[str] = None,
+    ) -> None:
+        # A list of storages and the translation mappers they are associated with.
+        # This list will only contain one storage and mappers for single storage entities.
+        # If there are more than one storage and mappers, a selector is required
+        self.__storages = storages
+        # A storage selector class to determine which to use in query plan
+        self.__selector = selector
+        # This is a set of query processors that have to be executed on the
+        # query after the storage selection but that are defined by the dataset.
```
[the remark this answers is not in the document]
Comment (enochtangg) [reply snuba/datasets/plans/storage_plan_builder_new.py:132]: Updated class names and comments.

**Comment by evanh** [comment]
@enochtangg That largely makes sense, thanks for clarifying.

This made me think of another question though: what's the long term plan with the mappers? Because those are actually tied to a storage (even though they are defined in the entity). I think some thought should go into where those mappers should live. If the storage pipeline has no mappers, how will that affect the queries that need to be written for that pipeline? How difficult would it be to have the mappers be applied after the clickhouse query is generated, instead of during the clickhouse generation?

**Review by evanh — APPROVED** [review]
OK looks good to me! Thanks @enochtangg

File: snuba/datasets/plans/storage_plan_builder_new.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        schema=relational_source.get_columns(),
+        allocation_policies=allocation_policies,
+        final=final,
+        sampling_rate=sampling_rate,
+        mandatory_conditions=relational_source.get_mandatory_conditions(),
+        storage_key=storage_key,
+    )
+
+
+class EntityProcessingExecutor:
```
Comment (kylemumma) [inline snuba/datasets/plans/storage_plan_builder_new.py:108]: should this class be in an entity-related file in order to separate it more from the storage-related concepts?

**Review by kylemumma — COMMENTED** [review]
Hi Enoch, this PR is very impressive, I can't believe you were able to make such large restructurings to the pipeline!

I wasn't able to follow everything you did 100% but here were my biggest questions:
1. It seemed to me that the entity stage contained a bit of storage logic, particularly the translation from entity to storage. I'm curious why you designed it this way (see my comments at the definitions of your QueryPipelineStages.)

2. The new pipeline you made seems significantly different in structure than old one. I got lost and was unable to follow the changes between the two, and only read through the new pipeline you wrote without comparing it to the old one. I am curious how / if you were able to ensure that all the same things are happening in the new pipeline as were happening in the old one.

File: snuba/pipeline/stages/query_processing.py [hunk 3]
```diff
@@ -43,3 +48,38 @@ def _query_runner(
...
+            pipe_input.data.query, pipe_input.query_settings
+        )
+        return clickhouse_query
+
+
+class StorageProcessingStage(
+    QueryPipelineStage[
+        ClickhouseQuery | CompositeQuery[Table],
+        ClickhouseQuery | CompositeQuery[Table],
+    ]
```
[the remark this answers is not in the document]
Comment (enochtangg) [reply snuba/pipeline/stages/query_processing.py:75]: See explanation above

File: snuba/datasets/plans/storage_plan_builder_new.py (hunk 2 above)
Comment (enochtangg) [reply snuba/datasets/plans/storage_plan_builder_new.py:108]: Yes, I agree! After figuring out the composite query pipeline, I plan to rename this file to something more appropriate. By keeping all the new classes here and with a filename suffixed with `-new` helps me differentiate between code that supports old and new pipeline.

File: snuba/pipeline/stages/query_processing.py [hunk 4]
```diff
@@ -43,3 +48,38 @@ def _query_runner(
         execution_pipeline.execute()
         assert _clickhouse_query is not None
         return _clickhouse_query
+
+
+class EntityProcessingStage(
+    QueryPipelineStage[Request, ClickhouseQuery | CompositeQuery[Table]]
```
[the remark this answers is not in the document]
Comment (kylemumma) [reply snuba/pipeline/stages/query_processing.py:54]: Ohh yeah that makes sense, thanks for the explanation! I wasn't thinking about the querying storages directly part...

**Pull request merged**